libcloud/common/zonomi.py

```diff
@@ -0,0 +1,150 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.You may obtain a copy of the License at
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from libcloud.common.base import XmlResponse
+from libcloud.common.base import ConnectionKey
+from libcloud.utils.py3 import b, PY3
+
+
+__all__ = [
+    'ZonomiException',
+    'ZonomiResponse',
+    'ZonomiConnection'
+]
+
+# Endpoint for Zonomi API.
+API_HOST = 'zonomi.com'
+
+SPECIAL_ERRORS = [
+    'Not found.',
+    'ERROR: This zone is already in your zone list.',
+    'Record not deleted.'
+]
+
+
+class ZonomiException(Exception):
+    def __init__(self, code, message):
+        self.code = code
+        self.message = message
+        self.args = (code, message)
+
+    def __str__(self):
+        return "%s %s" % (self.code, self.message)
+
+    def __repr__(self):
+        return "ZonomiException %s %s" % (self.code, self.message)
+
+
+class ZonomiResponse(XmlResponse):
+    errors = None
+    objects = None
+
+    def __init__(self, response, connection):
+        self.connection = connection
+        self.headers = dict(response.getheaders())
+        self.error = response.reason
+        self.status = response.status
+
+        # This attribute is used when using LoggingConnection
+        original_data = getattr(response, '_original_data', None)
```

Comment:
I noticed you duplicated this code from the base class in one other driver as well. I know there is a small difference at the bottom, but duplication like this is problematic (it makes code hard to maintain). On top of the duplication, this also break the implicit signature / contract of the `parse_body` method (it should return an object, not a tuple).

I will review this and merge this today, but after this is merged can you please open a new PR which reduces the duplication - you should re-use as much functionality from the parent class as possible.

One thing you could do is make `parse_body` a no-op and add a new `parse_body_and_errors` method and make constructor look something like this:

```python
def __init__(self, response, connection):
    super(Foo, self).__init__(response=response, connection=connection)

   self.objects, self.errors = self.parse_body_and_errors()
   if self.errors:
     raise self._make_excp(self.errors[0])
```